### PR TITLE
Feature/bloc favorites advanced details

### PR DESCRIPTION
**Pull request by StanislavPrime**

### CODE REVIEW

File: examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/details/blocs/puppy_details_bloc.dart [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    PuppyDetailsEvent event,
+  ) async* {
+    final puppy = event.puppy;
+    if (event.updateException.isEmpty) {
+      yield state.copyWith(
+          puppy: _puppy!.copyWith(isFavorite: puppy.isFavorite));
+    } else {
+      yield state.copyWith(
+          puppy: puppy.copyWith(isFavorite: !puppy.isFavorite));
+      await Future.delayed(const Duration(milliseconds: 400));
```
Comment (StanevPrime) [inline examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/details/blocs/puppy_details_bloc.dart:45]: @StanislavPrime, do we really need this artificial delay ?

File: examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/details/blocs/puppy_details_state.dart [hunk 2]
```diff
@@ -0,0 +1,45 @@
+part of 'puppy_details_bloc.dart';
+
+@immutable
+class PuppyDetailsState extends Equatable {
```
Comment (StanevPrime) [inline examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/details/blocs/puppy_details_state.dart:4]: @StanislavPrime, what is the reason for all properties being nullable? It looks like this causes a big impact on using that state in the UI layer, as every time it needs to be added a nullability check.

File: examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/details/ui_components/puppy_details.dart [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+              Padding(
+                padding: const EdgeInsets.all(16),
+                child: Column(
+                  crossAxisAlignment: CrossAxisAlignment.start,
+                  mainAxisAlignment: MainAxisAlignment.end,
+                  children: [
+                    const SizedBox(height: 10),
+                    BlocBuilder<PuppyDetailsBloc, PuppyDetailsState>(
+                        builder: (context, state) => Text(
+                              state.puppy?.name ?? puppy.name,
```
Comment (StanevPrime) [inline examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/details/ui_components/puppy_details.dart:47]: Think of a way of avoiding all of those nullability checks

File: examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/details/views/puppy_details_page.dart [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+class PuppyDetailsPage extends StatelessWidget {
+  const PuppyDetailsPage({
+    required Puppy puppy,
+    Key? key,
+  })  : _puppy = puppy,
+        super(key: key);
+
+  static Page page({required Puppy puppy}) => MaterialPage(
+        child: MultiBlocProvider(
```
Comment (StanevPrime) [inline examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/details/views/puppy_details_page.dart:20]: @StanislavPrime, what is the reason for registering two `MultiBlocProvider` ?

File: examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/favorites/blocs/favorite_puppies_bloc.dart [hunk 5]
```diff
@@ -82,23 +86,16 @@ class FavoritePuppiesBloc
...
-          error: e.toString(),
-        );
-        errorDisplayed = true;
-      }
       yield state.copyWith(
         favoritePuppies: state.favoritePuppies
             .manageList(isFavorite: !isFavorite, puppy: puppy),
+        error: e.toString(),
+      );
+      await Future.delayed(const Duration(milliseconds: 200));
```
Comment (StanevPrime) [inline examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/favorites/blocs/favorite_puppies_bloc.dart:95]: @StanislavPrime, is there a real need of adding an artificial delay here?

File: examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/search/views/search_page.dart [hunk 6]
```diff
@@ -44,28 +42,26 @@ class SearchPage extends StatelessWidget {
...
                 final item = state.searchedPuppies![index];
-                // print('PuppyListStatus  : ${state.status}');
                 return PuppyCard(
                   key: Key('${Keys.puppyCardNamePrefix}${item.id}'),
                   onVisible: (puppy) => context
                       .read<PuppiesExtraDetailsBloc>()
                       .add(FetchPuppyExtraDetailsEvent(puppy)),
                   puppy: item,
-                  onCardPressed: (puppy) {},
+                  onCardPressed: (puppy) {
```
Comment (StanevPrime) [inline examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/search/views/search_page.dart:51]: @StanislavPrime, here you can use a body expression

File: examples/favorites_advanced/bloc_favorites_advanced/pubspec.yaml [hunk 7]
```diff
@@ -37,6 +37,7 @@ dependencies:
   mockito: ^5.0.4
   declarative_animated_list: ^0.1.0-nullsafety.0
   clean_coverage: ^0.0.3
+  flow_builder: ^0.0.1
```
Comment (StanevPrime) [inline examples/favorites_advanced/bloc_favorites_advanced/pubspec.yaml:40]: @StanislavPrime, I believe there is a newer version

**Review by StanevPrime — CHANGES_REQUESTED** [review]

File: examples/favorites_advanced/bloc_favorites_advanced/pubspec.yaml (hunk 7 above)
Comment (StanislavPrime) [reply examples/favorites_advanced/bloc_favorites_advanced/pubspec.yaml:40]: I changed the version of  `flow_builder` to the newer 0.0.2

File: examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/search/views/search_page.dart (hunk 6 above)
Comment (StanislavPrime) [reply examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/search/views/search_page.dart:51]: I converted the function to an expression-bodied function: 
`onCardPressed: (puppy) =>
                    Navigator.of(context).push(PuppyFlow.route(puppy: puppy))`

File: examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/details/blocs/puppy_details_bloc.dart (hunk 1 above)
Comment (StanislavPrime) [reply examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/details/blocs/puppy_details_bloc.dart:45]: This delay is not needed and I removed it.

File: examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/favorites/blocs/favorite_puppies_bloc.dart (hunk 5 above)
Comment (StanislavPrime) [reply examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/favorites/blocs/favorite_puppies_bloc.dart:95]: When I yield 2 different states from the bloc one after another it seems that the UI is not rebuilding when there is no delay and I cannot see the expected change of the Icon. I found a similar issue here:
https://github.com/felangel/bloc/issues/459
The explanation is: You're simply yielding states faster than flutter can rebuild the UI.

File: examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/details/views/puppy_details_page.dart (hunk 4 above)
Comment (StanislavPrime) [reply examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/details/views/puppy_details_page.dart:20]: The application works the same way without the second `MultiBlocProvider`, so I removed it.

File: examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/details/blocs/puppy_details_state.dart (hunk 2 above)
Comment (StanislavPrime) [reply examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/details/blocs/puppy_details_state.dart:4]: I removed the nullable properties and left only a non-nullable `Puppy `property in the `PuppyDetailsState`

File: examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/details/ui_components/puppy_details.dart (hunk 3 above)
Comment (StanislavPrime) [reply examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/details/ui_components/puppy_details.dart]: I set a non-nullable `Puppy `in `PuppyDetailsState `and with that removed the nullability checks.

File: examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/favorites/blocs/favorite_puppies_bloc.dart (hunk 5 above)
Comment (StanislavPrime) [reply examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/favorites/blocs/favorite_puppies_bloc.dart]: I removed this delay from the method.

File: examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/details/blocs/puppy_details_bloc.dart (hunk 1 above)
Comment (StanislavPrime) [reply examples/favorites_advanced/bloc_favorites_advanced/lib/feature_puppy/details/blocs/puppy_details_bloc.dart]: I removed this delay from the method and added the overriden method `transformTransitions `with `transitions.interval(const Duration(milliseconds: 200));`

**Review by StanevPrime — APPROVED** [review]

**Pull request merged**